containers/payments/PayPal.js

```diff
@@ -1,43 +1,176 @@
-import React, { useEffect } from 'react';
+import React, { useEffect, useRef, useState } from 'react';
 import PropTypes from 'prop-types';
 import { c } from 'ttag';
-import { Alert, PrimaryButton, SmallButton, useApiResult, Price } from 'react-components';
-import { createPayPalPayment } from 'proton-shared/lib/api/payments';
-import { API_CUSTOM_ERROR_CODES } from 'proton-shared/lib/errors';
-import { parseURL } from 'proton-shared/lib/helpers/browser';
-import { MIN_PAYPAL_AMOUNT, MAX_PAYPAL_AMOUNT } from 'proton-shared/lib/constants';
+import { Alert, PrimaryButton, SmallButton, Price, useApi, useLoading } from 'react-components';
+import { createToken, getTokenStatus } from 'proton-shared/lib/api/payments';
+import {
+    MIN_PAYPAL_AMOUNT,
+    MAX_PAYPAL_AMOUNT,
+    PAYMENT_TOKEN_STATUS,
+    PAYMENT_METHOD_TYPES
+} from 'proton-shared/lib/constants';
+import { wait } from 'proton-shared/lib/helpers/promise';
+
+const {
+    STATUS_PENDING,
+    STATUS_CHARGEABLE,
+    STATUS_FAILED,
+    STATUS_CONSUMED,
+    STATUS_NOT_SUPPORTED
+} = PAYMENT_TOKEN_STATUS;
+
+const DELAY_PULLING = 5000;
+const DELAY_PAYPAL_TAB_LISTENER = 1000;
 
 const PayPal = ({ amount, currency, onPay, type }) => {
-    const { result = {}, loading, error = {}, request } = useApiResult(() => createPayPalPayment(amount, currency), []);
-    const { ApprovalURL } = result;
-    const reset = () => window.removeEventListener('message', receivePaypalMessage, false);
-    const handleClick = () => window.open(ApprovalURL, 'PayPal');
+    const [loading, withLoading] = useLoading();
+    const tabRef = useRef(null);
+    const timerRef = useRef(0);
+    const processRef = useRef(true);
+    const tokenRef = useRef('');
+    const api = useApi();
+    const [approvalURL, setAppovalURL] = useState('');
+    const [error, setError] = useState(null);
+
+    const I18N = {
```

Comment:
It would be better to move this outside the component, to avoid repeated calls on re-renders.

Reply:
This needs to be inside the component to refresh it when the language is changed.